### ISSUE TITLE
[Core feature] Exponential backoff retry

### ISSUE THREAD

**Issue by mkao006** [flyteorg/flyte#2333]
### Motivation: Why do you think this is important?

Sometimes there are downtimes for an upstream system which causes the workflow to fail, the workflow would succeed after retrying at a later time. 

The limitation with the current retry is the maximum count and lacking the flexibility to set the interval thus the retry is exhausted before the upstream system recover from downtime.

### Goal: What should the final outcome look like, ideally?

The retry will be performed on an increased interval [1m, 3m, 10m, 30m etc] up to the number of retries.

### Describe alternatives you've considered 

No alternatives are available at the moment.

### Are you sure this issue hasn't been raised already?

- [X] Yes

### Have you read the Code of Conduct?

- [X] Yes

**Comment by hamersaw** [comment]
Commenting to keep open.

**Comment by kumare3** [comment]
i really like this idea. 
I think from a flytekit pov, it should be implemented like so
```
# retries is simply implemented using a retry class, similar to container_image=ImageSpec or str
@task(retries=BackoffRetry(interval=timedelta(minutes=2), count=3))
def foo():
   ...
```

In flyteidl, we can add it to [RetryStrategy](https://github.com/flyteorg/flyte/blob/b6f35add1227c930e9208103235d89fe4b864b8e/flyteidl/protos/flyteidl/core/literals.proto#L179) which is already a class. 

In Flytepropeller we can add it to
[Propeller's struct](https://github.com/flyteorg/flyte/blob/b6f35add1227c930e9208103235d89fe4b864b8e/flytepropeller/pkg/apis/flyteworkflow/v1alpha1/nodes.go#L67)

and then to the computation logic [here](https://github.com/flyteorg/flyte/blob/b6f35add1227c930e9208103235d89fe4b864b8e/flytepropeller/pkg/compiler/transformers/k8s/utils.go#L21)

We can use the last_updated_at from the node_status to wait for the retry. 

This should not be too hard to do.
We can also implement exponential backoff

**Comment by kdhingra307** [comment]
@kumare3 will it be possible to add a feature where a user can provide the delay for retry? 

It is useful in case upstream system provides the delay mechanism

**Comment by kumare3** [comment]
Yup absolutely- this is indeed the idea with retry strategy and implementation should be easy. But someone has to be contribute

**Comment by kumare3** [comment]
@kdhingra307 join slack.flyte.org and we can collaborate

**Comment by kdhingra307** [comment]
@kumare3 sure let's connect on slack once

**Comment by kumare3** [comment]
As discussed if you just want the workflow to sleep / wait before task B, you can use Flytekit.sleep

Also nodes will be queued in case no resources are available
